addons/http_routing/models/ir_http.py

```diff
@@ -337,15 +341,184 @@ def get_nearest_lang(cls, lang_code):
+           endpoint.
+
+        2/ Use the URL as-is when the lang is not present in the URL and
+           that the default lang has been requested.
+
+        3/ Use the URL as-is saving the requested lang when the user is
+           a bot and that the lang is missing from the URL.
+
+        4/ Redirect the browser when the lang is missing from the URL
+           but another lang than the default one has been requested. The
+           requested lang is injected before the original path.
+
+        5) Use the url as-is when the lang is missing from the URL, that
+           another lang than the default one has been requested but that
+           it is forbidden to redirect (e.g. POST)
+
+        6/ Redirect the browser when the lang is present in the URL but
+           it is the default lang. The lang is removed from the original
+           URL.
+
+        7/ Redirect the browser when the lang present in the URL is an
+           alias of the preferred lang url code (e.g. fr_FR -> fr)
+
+        8/ Redirect the browser when the requested page is the homepage
+           but that there is a trailing slash.
+
+        9/ Rewrite the URL when the lang is present in the URL, that it
+           matches and that this lang is not the default one. The URL is
+           rewritten to remove the lang.
+
+        Note: The "requested lang" is (in order) either (1) the lang in
+              the URL or (2) the lang in the ``frontend_lang`` request
+              cookie or (3) the lang in the context or (4) the default
+              lang of the website.
+        """
+
+        # The URL has been rewritten already
+        if hasattr(request, 'is_frontend'):
+            return super()._match(path)
+
+        # See /1, match a non website endpoint
+        with contextlib.suppress(NotFound):
+            rule, args = super()._match(path)
+            routing = rule.endpoint.routing
+            request.is_frontend = routing.get('website', False)
+            request.is_frontend_multilang = request.is_frontend and routing.get('multilang', routing['type'] == 'http')
+            if not request.is_frontend:
+                return rule, args
+
+        _, url_lang_str, *rest = path.split('/', 2)
+        path_no_lang = '/' + (rest[0] if rest else '')
+        allow_redirect = request.httprequest.method != 'POST'
+
+        # There is no user on the environment yet but the following code
+        # requires one to set the lang on the request. Temporary grant
+        # the public user. Don't try it at home!
+        real_env = request.env
+        try:
+            request.registry['ir.http']._auth_method_public()  # it calls update_env
+            nearest_url_lang = cls.get_nearest_lang(request.env['res.lang']._lang_get_code(url_lang_str))
+            cookie_lang = cls.get_nearest_lang(request.httprequest.cookies.get('frontend_lang'))
+            context_lang = cls.get_nearest_lang(real_env.context.get('lang'))
```

Comment:
`real_env` seems only useful in to compute this case.

But in most of the case, we will never use it since you will have the `nearest_url_lang or cookie_lang `value first.

So can't we do it only `if not nearest_url_lang and not cookie_lang` to avoid to do _what you shouldn't do at home_?